### PR TITLE
Fix #1497 - Debug task for AppleTV

### LINKED ISSUE #1497: [Bug] Cannot launch debug configuration for Apple TV apps
# 🐛 What is the bug? How can we reproduce it?

1. Create a React Native TV project
2. Create a new `Debug iOS` launch VSCode configuration
3. The TV app is launched inside a new simulator, but the extension crashes with this message: 
4. After that, the debug task is no longer running

## Expected behavior

The app should launch inside a new simulator, which should be set in debug mode (wired to the running RN packager within VSCode).

## Debug output

<details>
<summary>Debug Console</summary>

```text

Launching the application

```
</details>

<details>
<summary>React Native output channel</summary>

```text

[Info] Packager is already running.

[Info] Prewarming bundle cache. This may take a while ...

[Info] About to get: http://localhost:8081/index.bundle?platform=ios

[Tue Feb 02 2021 14:33:47.840]  BUNDLE  ./index.js 

[Warning] The Bundle Cache was prewarmed.

[Info] Building and running application.

[Tue Feb 02 2021 14:33:59.912]  BUNDLE  ./index.js 

info Starting custom debugger by executing:, echo A debugger is not needed:  "/Users/redacted/git/myapp-tv"
[Tue Feb 02 2021 14:34:00.109]  BUNDLE  ./index.js 

[Error] Error: Error while executing command '/usr/libexec/PlistBuddy -c 'Print:CFBundleIdentifier' '/Users/redacted/Library/Developer/Xcode/DerivedData/MyTVApp-beepkmxzokbelpfrdgtacyzayayc/Build/Products/Debug-iphonesimulator/MyTVApp-tvOS.app/Info.plist'': Command failed: /usr/libexec/PlistBuddy -c 'Print:CFBundleIdentifier' '/Users/redacted/Library/Developer/Xcode/DerivedData/MyTVApp-beepkmxzokbelpfrdgtacyzayayc/Build/Products/Debug-iphonesimulator/MyTVApp-tvOS.app/Info.plist'
Print: Entry, ":CFBundleIdentifier", Does Not Exist

```
</details>

<details>
<summary>Developer Tools console</summary>

```text

<empty>

```
</details>

## Environment

Please tell us about your system and your project:

* `npx react-native doctor` or `npx expo doctor` if your project is using Expo:

```text
Common
 ✓ Node.js
 ✓ yarn
 ✓ Watchman - Used for watching changes in the filesystem when in development mode

Android
 ✖ JDK
   - Version found: 1.8.0_282
   - Version supported: >= 8
 ✖ Android Studio - Required for building and installing your app on Android
 ✓ Android SDK - Required for building and installing your app on Android
 ✖ ANDROID_HOME

iOS
 ✓ Xcode - Required for building and installing your app on iOS
 ✓ CocoaPods - Required for installing iOS dependencies
 ● ios-deploy - Required for installing your app on a physical device with the CLI

Errors:   3
Warnings: 1
```

* `envinfo`:

```text
  System:
    OS: macOS 11.1
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 633.24 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.15.4 - ~/.nvm/versions/node/v14.15.4/bin/node
    Yarn: 1.22.10 - ~/.nvm/versions/node/v14.15.4/bin/yarn
    npm: 6.14.10 - ~/.nvm/versions/node/v14.15.4/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  Managers:
    CocoaPods: 1.10.0 - /usr/local/bin/pod
    Homebrew: 2.7.7 - /usr/local/bin/brew
    pip3: 21.0.1 - /usr/local/bin/pip3
    RubyGems: 3.0.3 - /usr/bin/gem
  Utilities:
    Make: 3.81 - /usr/bin/make
    GCC: 4.2.1 - /usr/bin/gcc
    Git: 2.24.3 - /usr/bin/git
    Clang: 1200.0.32.29 - /usr/bin/clang
  Servers:
    Apache: 2.4.46 - /usr/sbin/apachectl
  SDKs:
    iOS SDK:
      Platforms: iOS 14.4, DriverKit 20.2, macOS 11.1, tvOS 14.3, watchOS 7.2
    Android SDK:
      API Levels: 28, 29, 30
      Build Tools: 28.0.3, 30.0.3
      System Images: android-29 | Android TV Intel x86 Atom, android-30 | Google APIs Intel x86 Atom
  IDEs:
    Nano: 2.0.6 - /usr/bin/nano
    Vim: 8.2 - /usr/bin/vim
    Xcode: 12.4/12D4e - /usr/bin/xcodebuild
  Languages:
    Bash: 3.2.57 - /bin/bash
    Java: 1.8.0_282 - /usr/bin/javac
    Perl: 5.28.2 - /usr/bin/perl
    PHP: 7.3.24 - /usr/bin/php
    Python: 2.7.16 - /usr/local/bin/python
    Python3: 3.9.1 - /usr/local/bin/python3
    Ruby: 2.6.3 - /usr/bin/ruby
  Databases:
    SQLite: 3.32.2 - /Users/redacted/Library/Android/sdk/platform-tools/sqlite3
  Browsers:
    Chrome: 88.0.4324.96
    Firefox: 85.0
    Safari: 14.0.2
```

* React Native Tools extension version: 1.3.0
* Expo SDK version (if applicable): n/a

### PR DESCRIPTION
**Pull request by micheleb**
Hi there 👋🏻 

I created a quick PR that fixes my issue. It's simply a fallback in case the configuration folder doesn't exist, so it should have minimal impact on users of the extension that don't work with AppleTV.

Thanks!

### CODE REVIEW

**Review by RedMickey — COMMENTED** [review]
Hi @micheleb and thanks a lot for your contribution!
I tested this PR on a new React Native tvOS project and it works fine, but if the project was built once for an iOS target, then the extension would fail with the original error, when a tvOS application is being tried to debug. This happens because after the build for an iOS simulator `/Users/user/Library/Developer/Xcode/DerivedData/TestTVApp-xxx/Build/Products/Debug-iphonesimulator` path already exists and [the conditions in the code](https://github.com/microsoft/vscode-react-native/pull/1498/files#diff-37ccebbac472ddd92f9d8c6935302cd78c97f59d14f1be003102ebb67501b3bfR76) work incorrectly.

I used the following debugging configuration to debug a tvOS app:
```
        {
            "name": "Debug iOS",
            "request": "launch",
            "type": "reactnative",
            "cwd": "${workspaceFolder}",
            "platform": "ios",
            "target": "Apple TV",
            "scheme": "TestTVApp-tvOS"
        }
```

Since `xxx-tvOS` [scheme](https://github.com/microsoft/vscode-react-native/blob/master/src/extension/ios/plistBuddy.ts#L40) is used to run a tvOS app, could it be better to check whether a scheme name contains `tvos` part to configure [sdk type](https://github.com/microsoft/vscode-react-native/blob/master/src/extension/ios/plistBuddy.ts#L57)?

Could you please also sign our Contributor License Agreement?

**Comment by micheleb** [comment]
Thanks!

> Since xxx-tvOS scheme is used to run a tvOS app, could it be better to check whether a scheme name contains tvos part to configure sdk type?

The thought of somebody developing for TVs and iPhones on the same codebase didn't cross my mind 😅 
But sure, sounds like a better idea, I'll do that instead.

> Could you please also sign our Contributor License Agreement?

I just switched jobs, so I'm checking with my new company what's their policy on these. Hopefully I'll be able to sign it very soon!

File: src/extension/ios/plistBuddy.ts [hunk 1]
```diff
@@ -54,13 +54,16 @@ export class PlistBuddy {
...
+            const deviceType = (scheme?.indexOf("-tvOS") ?? -1) > -1 ? "appletv" : "iphone";
+            const sdkType = `${deviceType}${sdkSuffix}`;
+
             let configurationFolder = path.join(productsFolder, `${configuration}-${sdkType}`);
             let executable = "";
             if (productName) {
                 executable = `${productName}.app`;
                 if (!fs.existsSync(path.join(configurationFolder, executable))) {
-                    const configurationData = this.getConfigurationData(
+                    let configurationData = this.getConfigurationData(
```
Comment (RedMickey) [inline src/extension/ios/plistBuddy.ts:66]:
```suggestion
                    const configurationData = this.getConfigurationData(
```

File: src/extension/ios/plistBuddy.ts [hunk 2]
```diff
@@ -69,12 +72,13 @@ export class PlistBuddy {
...
                         configurationFolder,
                     );
+
                     configurationFolder = configurationData.configurationFolder;
                 }
             } else {
                 const executableList = this.findExecutable(configurationFolder);
                 if (!executableList.length) {
-                    const configurationData = this.getConfigurationData(
+                    let configurationData = this.getConfigurationData(
```
Comment (RedMickey) [inline src/extension/ios/plistBuddy.ts:81]:
```suggestion
                    const configurationData = this.getConfigurationData(
```

File: src/extension/ios/plistBuddy.ts [hunk 3]
```diff
@@ -54,13 +54,16 @@ export class PlistBuddy {
             } else {
                 productsFolder = path.join(iosProjectRoot, "build", "Build", "Products");
             }
-            const sdkType = simulator ? "iphonesimulator" : "iphoneos";
+            const sdkSuffix = simulator ? "simulator" : "os";
+            const deviceType = (scheme?.indexOf("-tvOS") ?? -1) > -1 ? "appletv" : "iphone";
```
Comment (RedMickey) [inline src/extension/ios/plistBuddy.ts:58]:
```suggestion
            const deviceType = 
                (scheme?.toLowerCase().indexOf("tvos") ?? -1) > -1 ? "appletv" : "iphone";
```

**Review by RedMickey — COMMENTED** [review]
Hi @micheleb, I tested AppleTV debugging feature and it works fine, thank you! I left few little comments in the PR.
I also suppose that implementation of `sdkType` selection algorithm can be moved into a separate function.
Also it would be great if this feature will be covered with some tests. For example, [here](https://github.com/microsoft/vscode-react-native/blob/master/test/extension/ios/plistBuddy.test.ts).

File: src/extension/ios/plistBuddy.ts (hunk 3 above)
Comment (micheleb) [reply src/extension/ios/plistBuddy.ts:58]: I was under the impression that those suffixes were set automatically when creating the project, so the original version would be a tiny bit more robust (it works even if the project name contains the `tvos` string). I'm not an Xcode expert though, so I'll go with your suggestion. Just wanted to share the rationale behind the choice.

**Review by RedMickey — COMMENTED** [review]
Could you please run `gulp format:prettier` command in the terminal in the project root folder to fix code style issues?
It requires to have [`gulp` package installed globally](https://gulpjs.com/docs/en/getting-started/quick-start#install-the-gulp-command-line-utility) to run the command.

**Comment by micheleb** [comment]
done, there was a trailing whitespace introduced by accepting one of the proposed changes 😬 

As for the test and extracting the function, sure, I'll add that when I have time.

**Comment by micheleb** [comment]
Hey there, I extracted the `sdkType` detection logic into a dedicated function, and integrated the new logic in the existing `getBundleId` test in `plistBuddy.test.ts` + added another test for an appleTV scheme. Hope that's enough :)  Thanks!

**Review by RedMickey — APPROVED** [review]
Thanks a lot for your work! This fix will be available in the next version of the extension.

**Pull request merged**